NYPL/fedora_ingest_rails
# Feature/dr 2342 delete parents if empty

## Jira Tickets ##
[Ensure deleted containers in collections are removed from Repo API index](https://jira.nypl.org/browse/DR-2342)

## Things Done ##
- Created new method in the repo solr client to check if an item's previous parents would be empty upon update of this item's current solr doc, or if they are already empty. If either condition is met, the parent will be deleted from the index via its uuid. 

## How to Test ##
- Wish this was going to be easier. Here we go. 

Here are step by step instructions (with a tip of the hat to Jeff Croyle for laying the groundwork for these steps):

### Local Development ###
- rebuild your fedora ingest application to add recent dependencies: `docker-compose build`
- Configure your application to use some QA "stuff" (you'll want to revert these changes when you're done testing)
```
# these connections are not used locally _or_ in qa but can help prove out that fedora ingest test mode works without them
FEDORA_USERNAME=fedoraAdmin #qa 
FEDORA_PASSWORD=COWwNBsuHUWmliEDNZUpvIRGQVXhKprTK7hcxMRLkw8ZctoSUj #qa 
FEDORA_URL=http://qa01-fmaster.repo.nypl.org:8080/fedora #qa 

IMAGE_FILESTORE_DATABASE_HOST=prod.mysql.nypl.org #qa 
IMAGE_FILESTORE_DATABASE_NAME=archive #qa 
IMAGE_FILESTORE_DATABASE_USER=digital_read #qa 
IMAGE_FILESTORE_DATABASE_PASSWORD=d1g1t@l #qa 

AMI_FILESTORE_DATABASE_HOST='' #qa 
AMI_FILESTORE_DATABASE_NAME='' #qa 
AMI_FILESTORE_DATABASE_USER='' #qa
AMI_FILESTORE_DATABASE_PASSWORD=mysqlpassword #qa

MMS_URL=https://qa-metadata.nypl.org:443

REPO_SOLR_URL=http://10.225.131.109:8983/solr/repoapi
```
- bring up your app with `docker-compose up` and in another tab do `docker-compose run webapp bash`
- You should now be able to do `RAILS_ENV=test bundle exec rspec` and the rspecs will run without having to do anything else
- in your webapp session, cd to /logs and `tail -f development.log`
- Create two fake collections in the index by posting docs with just a fake uuid to solr (item uuids should be real uuids):

curl -X POST -H "Content-Type: application/json" -d '[
  {
    "uuid": "fake-soon-to-be-empty-collection-uuid"
  }
]' 'http://10.225.131.109:8983/solr/repoapi/update?commit=true'

curl -X POST -H "Content-Type: application/json" -d '[
  {
    "uuid": "fake-not-empty-collection-uuid"
  }
]' 'http://10.225.131.109:8983/solr/repoapi/update?commit=true'


curl -X POST -H "Content-Type: application/json" -d '[
  {
    "uuid": "6c8b2d20-c623-012f-39e8-58d385a7bc34",
    "parentUUID": ["fake-soon-to-be-empty-collection-uuid"]
  }
]' 'http://10.225.131.109:8983/solr/repoapi/update?commit=true'

curl -X POST -H "Content-Type: application/json" -d '[
  {
    "uuid": "6cd03ec0-c623-012f-90a6-58d385a7bc34",
    "parentUUID": ["fake-not-empty-collection-uuid"]
  }
]' 'http://10.225.131.109:8983/solr/repoapi/update?commit=true'

curl -X POST -H "Content-Type: application/json" -d '[
  {
    "uuid": "5b760b90-c226-0130-8b15-58d385a7b928",
    "parentUUID": ["fake-not-empty-collection-uuid"]
  }
]' 'http://10.225.131.109:8983/solr/repoapi/update?commit=true'

- Open postman and do a post to http://localhost:3000/ingest_requests with these properties:
    - Add Header `Content-Type: application/json`
    - Add Body -> raw 
```
{
    "uuids": ["6c8b2d20-c623-012f-39e8-58d385a7bc34"],
    "test_mode": true
}
```
- Post the post. You should get back a 201 Created with an empty body
- Check the development logs. There should be no errors and the Delayed Worker should have worked off the existing job successfully.
- Check out `http://10.225.131.109:8983/solr/repoapi/select?q=((%2Buuid:(%2B%2214b44ac0-c08a-0134-67c6-00505686a51c%22)))` (or whatever your doc is). It should have an updated `dateIndexed_s` value based on the time you posted (e.g. `"firstIndexed_s":["2023-04-14T20:46:02+00:00"]`)
   - Check out http://10.225.131.109:8983/solr/repoapi/select?q=((%2Buuid:(%2B%22fake-soon-to-be-empty-collection-uuid%22))). It should be gone.
   - Check out http://10.225.131.109:8983/solr/repoapi/select?q=((%2Buuid:(%2B%22fake-not-empty-collection-uuid%22))). It should NOT be gone.
   -Delete the last bogus collection via
  
  curl -X POST -H "Content-Type: application/json" -d '{
  "delete": {
    "query": "uuid:fake-not-empty-collection-uuid"
  }
}' 'http://10.225.131.109:8983/solr/repoapi/update?commit=true'

### QA ###
- Acccessinate at http://nypl-digital-dev.nypl.org/index.php
- Open postman and do a post to https://qa-fedora-ingestor.nypl.org/ingest_requests with these properties:
    - Add Header `Content-Type: application/json`
    - Add Body -> raw 
```
{
    "uuids": ["14b44ac0-c08a-0134-67c6-00505686a51c","6cd03ec0-c623-012f-90a6-58d385a7bc34"],
    "test_mode": true
}
```
   - Check out http://10.225.131.109:8983/solr/repoapi/select?q=((%2Buuid:(%2B%22fake-soon-to-be-empty-collection-uuid%22))). It should be gone.
   - Check out http://10.225.131.109:8983/solr/repoapi/select?q=((%2Buuid:(%2B%22fake-not-empty-collection-uuid%22))). It should NOT be gone.
   - Rinse and repeat as many times as you want, but at the end you probably want to trigger an update for the third legit UUID via... 
   - Open postman and do a post to https://qa-fedora-ingestor.nypl.org/ingest_requests with these properties:
    - Add Header `Content-Type: application/json`
    - Add Body -> raw 
```
{
    "uuids": ["5b760b90-c226-0130-8b15-58d385a7b928"],
    "test_mode": true
}
```
  And then delete the last bogus collection via
  
  curl -X POST -H "Content-Type: application/json" -d '{
  "delete": {
    "query": "uuid:fake-not-empty-collection-uuid"
  }
}' 'http://10.225.131.109:8983/solr/repoapi/update?commit=true'
